### PR TITLE
Implementation of true asynchronous signaling between processes

### PR DESCRIPTION
**Pull request by rickard-green**
Communication between Erlang processes has conceptually always been
performed through asynchronous signaling. The runtime system
implementation has however previously preformed most operation
synchronously. In a system with only one true thread of execution, this
is not problematic (often the opposite). In a system with multiple threads
of execution (as current runtime system implementation with SMP support)
it becomes problematic. This since it often involves locking of structures
when updating them which in turn cause resource contention. Utilizing
true asynchronous communication often avoids these resource contention
issues.

The case that triggered this change was contention on the link lock due
to frequent updates of the monitor trees during communication with a
frequently used server. The signal order delivery guarantees of the
language makes it hard to change the implementation of only some signals
to use true asynchronous signaling. Therefore the implementations
of (almost) all signals have been changed.

Currently the following signals have been implemented as true
asynchronous signals:
- Message signals
- Exit signals
- Monitor signals
- Demonitor signals
- Monitor triggered signals (DOWN, CHANGE, etc)
- Link signals
- Unlink signals
- Group leader signals

All of the above already defined as asynchronous signals in the
language. The implementation of messages signals was quite
asynchronous to begin with, but had quite strict delivery constraints
due to the ordering guarantees of signals between a pair of processes.

The previously used message queue partitioned into two halves has been
replaced by a more general signal queue partitioned into three parts
that service all kinds of signals. More details regarding the signal
queue can be found in comments in the erl_proc_sig_queue.h file.

The monitor and link implementations have also been completely replaced
in order to fit the new asynchronous signaling implementation as good
as possible. More details regarding the new monitor and link
implementations can be found in the erl_monitor_link.h file.

### CODE REVIEW

**Comment by rickard-green** [comment]
@kostis @margnus1 

A couple of things regarding receive and hipe are effected by this PR.

* hipe_check_get_msg() needs improvements. It should bump reductions according to what erts_proc_sig_receive_helper() returns, and also yield if necessary. See loop_rec instruction in msg_instrs.tab. As it is now it may hog the scheduler for a very long time.
* The message queue optimization lost some of its effectiveness due to the layout of the new signal queue. Hipe code will now set save_last to the end of inner queue which is safe, but might cause unnecessary scanning of messages in the middle and outer queue. It can be fixed by introducing some new hipe bifs I guess. You typically want to use the ERTS_RECV_MARK_{SAVE|SET|CLEAR} macros (defined in erl_message.h; see also usage in msg_instrs.tab).

File: erts/emulator/beam/erl_proc_sig_queue.h [hunk 1]
```diff
@@ -0,0 +1,661 @@
...
+ *              - Persistent monitor message
+ *              - Link
+ *              - Unlink
+ *              - Group leader
+ *              - Trace change
+ *
+ *              The signal queue consists of three parts:
+ *              - Outer queue (sig_inq field in process struct)
+ *              - Middle queue (sig_qs field in process struct)
+ *              - Inner queue (sig_qs field in process struct)
```
Comment (fenollp) [inline erts/emulator/beam/erl_proc_sig_queue.h]: You named the same field twice here. Shouldn't they be different fields?

**Comment by fenollp** [comment]
As I understand it this patch makes the receiving messages implementation effectively lockfree? Did you do any more or less representative benchmarking on for instance the monitoring example you cite?
Thanks!

**Comment by rickard-green** [comment]
@fenollp 

No not lockfree. The outer part of the signal queue is still protected by the process-msgq-lock. Inserting a signal compared to modifying a potentially very large search tree which might need rebalancing is however a much quicker operation so the msgq locking has much less impact.

The gain very much depends on the amount of cpu:s and schedulers. As an example, a benchmark with many clients making a large amount of requests to a single gen_server on a dual 8-core AMD opteron had a speedup of 4-6 times. We've seen heavy contention on link-locks in the "single server frequently utilized by lots of processes" scenario in real systems, so this is more or less what we set out to improve.

File: erts/emulator/beam/erl_proc_sig_queue.h (hunk 1 above)
Comment (rickard-green) [reply erts/emulator/beam/erl_proc_sig_queue.h]: No this is correct. Both the middle and inner queues are private to the process and also share some parts which is the reason why they are in the same field. See the ErtsSignalPrivQueues type in erl_message.h for details.

**Pull request merged**